pkg/policy/rule.go

```diff
@@ -94,32 +112,63 @@ func mergeL4Port(ctx *SearchContext, r api.PortRule, p api.PortProtocol, dir str
 		ctx.PolicyTrace("   Merge conflict: mismatching redirect ports %d/%d\n", l4Filter.L7RedirectPort, v.L7RedirectPort)
 		return 0, fmt.Errorf("Cannot merge conflicting redirect ports (%d/%d)", l4Filter.L7RedirectPort, v.L7RedirectPort)
 	}
-	switch {
-	case len(l4Filter.L7Rules.HTTP) > 0:
-		if len(v.L7Rules.Kafka) > 0 {
-			ctx.PolicyTrace("   Merge conflict: mismatching L7 rule types.\n")
-			return 0, fmt.Errorf("Cannot merge conflicting L7 rule types")
+
+	if adjustL4PolicyIfNeeded(fromEndpoints, &v) && len(r.Rules.HTTP) == 0 {
+		// skip this policy as it is already covered and it does not contain L7 rules
+		return 1, nil
+	}
+
+	// if (1) the existing rule did not have a wildcard endpoint
+	// AND (2) the new rule does not have explicit fromEndpoints
+	// THEN we need to copy all existing L7 rules to the wildcard endpoint
+	if _, ok := v.L7RulesPerEp[WildcardEndpointSelector]; !ok && len(fromEndpoints) == 0 {
+		wildcardEp := api.L7Rules{}
+		for _, existingL7Rules := range v.L7RulesPerEp {
+			wildcardEp.HTTP = append(wildcardEp.HTTP, existingL7Rules.HTTP...)
+			wildcardEp.Kafka = append(wildcardEp.Kafka, existingL7Rules.Kafka...)
 		}
-		v.L7Rules.HTTP = append(v.L7Rules.HTTP, l4Filter.L7Rules.HTTP...)
-	case len(l4Filter.L7Rules.Kafka) > 0:
-		if len(v.L7Rules.HTTP) > 0 {
-			ctx.PolicyTrace("   Merge conflict: mismatching L7 rule types.\n")
-			return 0, fmt.Errorf("Cannot merge conflicting L7 rule types")
+		v.L7RulesPerEp[WildcardEndpointSelector] = wildcardEp
+	}
+
+	for hash, newL7Rules := range l4Filter.L7RulesPerEp {
+		if ep, ok := v.L7RulesPerEp[hash]; ok {
+			switch {
+			case len(l4Filter.L7RulesPerEp[hash].HTTP) > 0:
```

Comment:
This is just something to note - how will we handle these types of 'collisions' in the future as more and more protocols are supported?

Reply:
Definitely something we should look at addressing by the time we consider supporting a third protocol.